File: cocos2d-html5_i/IronCity/src/layer/GameSceneOverLayer.js
/**
 * Created with JetBrains WebStorm.
 * User: cocos
 * Date: 13-10-23
 * Time: 下午3:36
 * To change this template use File | Settings | File Templates.
 */

var GameSceneOverLayer = cc.UILayer.extend({
    finalScore:null,
    parentScene:null,
    init:function(){
        if( cc.UILayer.prototype.init.call(this) ){
            this.parentScene = GameScene.getScene();
            this.addWidget(cc.UIHelper.getInstance().createWidgetFromJsonFile(Json_GameSceneOverLayer_1));

            var playAgainBtn        = this.getWidgetByName("playAgain");
            var monsterGroundAmount = this.getWidgetByName("monsterGroundLabel");
            var monsterSkyAmount    = this.getWidgetByName("monsterSkyLabel");
            var distanceScore      = this.getWidgetByName("distanceScore");
            this.finalScore = this.getWidgetByName("finalScore");

            playAgainBtn.setTouchEnabled(true);
            playAgainBtn.addTouchEventListener(this, this.playAgainBtnCallback);
            console.log("playAgainBtn: ", playAgainBtn);

            var monsterGroundCount= this.parentScene.playLayer.getMonsterGroundAmount();
            monsterGroundAmount.setText(monsterGroundCount);

            var monsterSkyCount= this.parentScene.playLayer.getMonsterSkyAmount();
            monsterSkyAmount.setText(monsterSkyCount);

            distanceScore.setStringValue(this.parentScene.menuLayer.getDistanceScore());
            this.calculateFinalScore(monsterGroundCount*88 , monsterSkyCount*66 , this.parentScene.menuLayer.getDistanceScore());

            //this.setTouchEnabled(true);
            //this.setTouchMode(cc.TouchMode);
            return true;
        }
        return false;
    },
//    onTouchBegan:function(touch){
//        console.log("touch", touch);
//    },
//    onTouchesBegan:function(touch){
//        console.log("touches", touch);
//    },
    calculateFinalScore:function(monsterSkyAmountValue, monsterGroundAmountValue, distanceScoreValue){
        var distanceScore = 0;
        distanceScore = distanceScoreValue;

        var score = monsterSkyAmountValue + monsterGroundAmountValue + distanceScore * 3;
        this.finalScore.setStringValue(score);
    },
    playAgainBtnCallback:function(pSender, type){
        console.log("playAgainBtnCallback: ", pSender, type);
        if(cc.TouchEventType.BEGAN == type){
            var againScene = new GameScene();
            againScene.init();
            var playAgainTransition =  cc.TransitionFade.create(0.5, againScene, cc.WHITE);
            cc.Director.getInstance().replaceScene(playAgainTransition);
        }
    }
});

File: cocos2d-x-2.2.0/projects/Ironman/Classes/GameSceneOverLayer.cpp
//
//  GameSceneOverLayer.cpp
//  Ironman
//
//  Created by Ken on 13-10-15.
//
//

#include "GameSceneOverLayer.h"
#include "stdlib.h"

bool GameSceneOverLayer::init()
{
    if(UILayer::init()){
        
		parentScene = GameScene::shareGameScene();
        this->addWidget(dynamic_cast<Layout*>(CCUIHELPER->createWidgetFromJsonFile("iphone/GameSceneOverLayer_1.json")));
        
        UIButton*    playAgainBtn        = dynamic_cast<UIButton*>(this->getWidgetByName("playAgain"));
		UILabel* monsterGroundAmount = dynamic_cast<UILabel*>(this->getWidgetByName("monsterGroundLabel"));
        UILabel* monsterSkyAmount    = dynamic_cast<UILabel*>(this->getWidgetByName("monsterSkyLabel"));
        UILabelAtlas* distanceScore      = dynamic_cast<UILabelAtlas*>(this->getWidgetByName("distanceScore"));
        finalScore = dynamic_cast<UILabelAtlas*>(this->getWidgetByName("finalScore"));
        
        playAgainBtn->addTouchEventListener(this, toucheventselector(GameSceneOverLayer::playAgainBtnCallback));

		int monsterGroundCount= parentScene->playLayer->getMonsterGroundAmount();
		ostringstream oss1;
        oss1<<monsterGroundCount;
	    string str1 = oss1.str();
	    const char * charStr1= str1.c_str();
        monsterGroundAmount->setText(charStr1);

		int monsterSkyCount= parentScene->playLayer->getMonsterSkyAmount();
		ostringstream oss2;
        oss2<<monsterSkyCount;
	    string str2= oss2.str();
	    const char * charStr2= str2.c_str();
        monsterSkyAmount->setText(charStr2);

        distanceScore->setStringValue(parentScene->menuLayer->getDistanceScore());
		this->calculateFinalScore(monsterGroundCount*88 , monsterSkyCount*66 , parentScene->menuLayer->getDistanceScore());
	
        return true;
    }
	
    return false;
}

void GameSceneOverLayer::calculateFinalScore(int monsterSkyAmountValue,int monsterGroundAmountValue, const char* distanceScoreValue)
{
	int distanceScore = 0;
	distanceScore = atoi(distanceScoreValue);
	
    int score = monsterSkyAmountValue + monsterGroundAmountValue + distanceScore * 3;
	
	ostringstream oss;
	oss<<score;
	string str= oss.str();
	const char * charStr2= str.c_str();
    
    finalScore->setStringValue(charStr2);
}

void GameSceneOverLayer::playAgainBtnCallback(CCObject *pSender, TouchEventType type)
{
	if(TOUCH_EVENT_BEGAN == type){
		
	    GameScene* againScene = GameScene::newGameScene();
        CCTransitionFade* playAgainTransition =  CCTransitionFade::create(0.5, againScene, ccWHITE);
        CCDirector::sharedDirector()->replaceScene(playAgainTransition);
	}
}
